pandas/tests/groupby/test_groupby.py

```diff
@@ -2174,3 +2174,20 @@ def test_groupby_mean_duplicate_index(rand_series_with_duplicate_datetimeindex):
     result = dups.groupby(level=0).mean()
     expected = dups.groupby(dups.index).mean()
     tm.assert_series_equal(result, expected)
+
+
+def test_groupby_categorical_crossproduct():
+    n = 10000
+    df = DataFrame({
+        "A": Categorical(np.arange(n), np.arange(n)),
+    })
+    for i in range(ord('B'), ord('L')):
+        v = np.tile(np.arange(25), n // 25)
+        df[chr(i)] = v
+
+    message = r"Group by product space too large to allocate arrays.*"
+    with pytest.raises(ValueError, match=message):
+        g = df.groupby(list(df.columns))
+        g.B.sum()
```

Comment:
If this doesn't raise a ValueError because of some changes to the code, then I worry what it might to do the CI. Maybe it'd be a good idea to add a timeout context manager to pandas._testing.contexts:

```
@contextmanager
def timeout(seconds):
    def timeout_handler(signum, frame):
        raise RuntimeError

    orig_handler = signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(seconds)
    try:
        yield
    finally:
        signal.alarm(0)
        signal.signal(signal.SIGALRM, orig_handler)
```

(Stolen from [SO](https://stackoverflow.com/questions/15018519/python-timeout-context-manager-with-threads))

Then you can do:

```
g = df.groupby(list(df.columns))
with pytest.raises(ValueError, match=message):
    with timeout(1):
        g.B.sum()
```

1 second should be well more than enough, it takes half a millisecond on my machine.